src/neo/SmartContract/Native/Oracle/OracleContract.cs

```diff
@@ -0,0 +1,230 @@
+        {
+            StoreView snapshot = engine.Snapshot;
+            ECPoint consignorPubKey = args[0].GetSpan().AsSerializable<ECPoint>();
+            ECPoint consigneePubKey = args[1].GetSpan().AsSerializable<ECPoint>();
+            ECPoint[] oraclePubKeys = NativeContract.NEO.GetValidators(snapshot);
+            if (!oraclePubKeys.Contains(consignorPubKey)) return false;
+            UInt160 account = Contract.CreateSignatureRedeemScript(consignorPubKey).ToScriptHash();
+            if (!InteropService.Runtime.CheckWitnessInternal(engine, account)) return false;
+            StorageKey key = CreateStorageKey(Prefix_Validator, consignorPubKey);
+            StorageItem item = snapshot.Storages.GetAndChange(key, () => new StorageItem());
+            item.Value = consigneePubKey.ToArray();
+
+            byte[] prefixKey = StorageKey.CreateSearchPrefix(Id, new[] { Prefix_Validator });
+            List<ECPoint> delegatedOracleValidators = snapshot.Storages.Find(prefixKey).Select(p =>
+              (
+                  p.Key.Key.AsSerializable<ECPoint>(1)
+              )).ToList();
+            foreach (var validator in delegatedOracleValidators)
+            {
+                if (!oraclePubKeys.Contains(validator))
+                {
+                    snapshot.Storages.Delete(CreateStorageKey(Prefix_Validator, validator));
+                }
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Get current authorized Oracle validator.
+        /// </summary>
+        /// <param name="engine">VM</param>
+        /// <param name="args">Parameter Array</param>
+        /// <returns>Authorized Oracle validator</returns>
+        [ContractMethod(0_01000000, ContractParameterType.Array)]
+        private StackItem GetOracleValidators(ApplicationEngine engine, Array args)
+        {
+            return new Array(engine.ReferenceCounter, GetOracleValidators(engine.Snapshot).Select(p => (StackItem)p.ToArray()));
+        }
+
+        /// <summary>
+        /// Get current authorized Oracle validator
+        /// </summary>
+        /// <param name="snapshot">snapshot</param>
+        /// <returns>Authorized Oracle validator</returns>
+        public ECPoint[] GetOracleValidators(StoreView snapshot)
+        {
+            ECPoint[] cnPubKeys = NativeContract.NEO.GetValidators(snapshot);
+            ECPoint[] oraclePubKeys = new ECPoint[cnPubKeys.Length];
+            System.Array.Copy(cnPubKeys, oraclePubKeys, cnPubKeys.Length);
+            for (int index = 0; index < oraclePubKeys.Length; index++)
+            {
+                var oraclePubKey = oraclePubKeys[index];
+                StorageKey key = CreateStorageKey(Prefix_Validator, oraclePubKey);
+                ECPoint delegatePubKey = snapshot.Storages.TryGet(key)?.Value.AsSerializable<ECPoint>();
+                if (delegatePubKey != null) { oraclePubKeys[index] = delegatePubKey; }
+            }
+            return oraclePubKeys;
```

Comment:
Shall we remove the `null` entries?

Reply:
No, the logic here is to return the public key of the delegate node if  it has been set. Otherwise,continue to return the public key of the CN node. So there is no need to delete